Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.cpp

```diff
@@ -437,7 +437,82 @@ LCMSProfilePtr JPEGXLImageDecoder::tryDecodeICCColorProfile()
+{
+    if (!m_iccTransform)
+        return;
+
+    cmsDoTransform(m_iccTransform.get(), inputBuffer, outputBuffer, numberOfPixels);
+}
+
+#elif USE(CG)
+
+void JPEGXLImageDecoder::clearColorTransform()
+{
+    m_profile = nullptr;
+}
+
+void JPEGXLImageDecoder::prepareColorTransform()
+{
+    if (m_profile || !m_basicInfo || m_basicInfo->num_extra_channels > 1 || m_basicInfo->exponent_bits_per_sample || m_basicInfo->alpha_exponent_bits)
+        return;
+
+    m_profile = tryDecodeICCColorProfile();
+    // TODO(bugs.webkit.org/show_bug.cgi?id=234222): We should try to use encoded color profile if ICC profile is not available.
+}
+
+RetainPtr<CGColorSpaceRef> JPEGXLImageDecoder::tryDecodeICCColorProfile()
+{
+    size_t profileSize;
+    if (JxlDecoderGetICCProfileSize(m_decoder.get(), &s_pixelFormat, JXL_COLOR_PROFILE_TARGET_DATA, &profileSize) != JXL_DEC_SUCCESS)
+        return nullptr;
+
+    Vector<uint8_t> profileData(profileSize);
+    if (JxlDecoderGetColorAsICCProfile(m_decoder.get(), &s_pixelFormat, JXL_COLOR_PROFILE_TARGET_DATA, profileData.data(), profileData.size()) != JXL_DEC_SUCCESS)
+        return nullptr;
+
+    return adoptCF(CGColorSpaceCreateWithICCData(adoptCF(CFDataCreate(kCFAllocatorDefault, profileData.data(), profileSize)).get()));
+}
+
+void JPEGXLImageDecoder::maybePerformColorSpaceConversion(void* inputBuffer, void* outputBuffer, unsigned numberOfPixels)
+{
+    if (!m_profile)
+        return;
+
+    // https://developer.apple.com/documentation/accelerate/1399134-vimageconvert_anytoany?language=objc
+    // "The destination buffer may only alias the srcs buffers if vImageConverter_MustOperateOutOfPlace returns 0 and the respective scanlines of the aliasing buffers start at the same address."
+    // convertImagePixels() doesn't have any logic for this, so we can just pessimize here and assume aliasing is illegal.
+    std::unique_ptr<uint8_t[]> intermediateBuffer(new uint8_t[4  * numberOfPixels]);
+    auto alphaFormat = m_premultiplyAlpha ? AlphaPremultiplication::Premultiplied : AlphaPremultiplication::Unpremultiplied;
+    ConstPixelBufferConversionView source {
+        .format = {
+            .alphaFormat = alphaFormat,
+            .pixelFormat = PixelFormat::BGRA8,
+            .colorSpace = DestinationColorSpace(m_profile.get()),
+        },
+        .bytesPerRow = static_cast<unsigned>(4 * size().width()),
+        .rows = static_cast<const uint8_t*>(inputBuffer),
+    };
+    PixelBufferConversionView destination {
+        .format = {
+            .alphaFormat = alphaFormat,
+            .pixelFormat = PixelFormat::BGRA8,
+            .colorSpace = DestinationColorSpace::SRGB(),
+        },
+        .bytesPerRow = static_cast<unsigned>(4 * size().width()),
+        .rows = intermediateBuffer.get(),
+    };
+    convertImagePixels(source, destination, { static_cast<int>(numberOfPixels), 1 });
```

Comment:
Feels a little odd to be pretending this is a _large_number_ x 1 dimension image. `bytesPerRow` is set to the right thing. I think we should just pass in `size()` here. Unless you are concerned about a mismatch between `size()` and `numberOfPixels`? (If so, you could validate that they match.)

Reply:
We can't - that's not the design of AppleJPEGXL. This code is all called from a callback that we give to AppleJPEGXL, and AppleJPEGXL will call the callback whenever it wants to emit a bunch of pixels in a row. We don't have the whole image here, we only have whatever chunk AppleJPEGXL decides to give us (which is guaranteed to be one-dimensional)

Reply:
OK. Kind of makes me feel like there should be a `convertImagePixels` overload that takes a number of pixels, instead of using `IntSize`. Can you confirm that `convertImagePixels` will do the right thing in terms of stride when we pass in an `IntSize` whose width does not correspond to the `bytesPerRow` we specify?